### DISCUSSION TITLE
Nextjs 13 router changes but show outdated data (caching / service worker / stale data) please help...

### DISCUSSION THREAD

**Discussion by yeonhwan** [vercel/next.js#52024] [Help] [answered]
### Summary

I am developing application with Next 13 (^ 13.2.4 / pages directory ) and T3 stack (tRPC, Prisma) now.
I made the 4 routes(myapp/home, myapp/settings ... etc) which use getServerSideProps to generate initial page layouts and hydrate.
Those 4 routes are sharing one information source (from database) and can be manipulated on each routes.

The problem is ,at some point, my application uses stale data to generate pages which is cached to next-data previously, not from getServerSideProps that is freshly fetched. (getServerSideProps runs just fine but it does not pass down the props to the _app components)

I vaguely suspect this is happening since I built the application with next-pwa package and tweaked some next.config.

However, even though I disable the next-pwa and restore my next config back, my next application keep getting the data from 'next-data' which is triggered by router.js

How can I sync the cached data with the return of getServerSideProps function or at least disable the caching behavior?

I searched for answers hours and hours but nothing is in use.... (Service Workers, next-pwa runtime caching, unstableSkipClientCache of router.push, setting headers in getServerSideProps)

plz help folks... I am completely lost and about the throw away the project right now :(

![스크린샷 2023-06-30 오후 5 34 34](https://github.com/vercel/next.js/assets/81786662/f077f4a5-66b3-47b7-a3c1-ae8a648b1a56)
![스크린샷 2023-06-30 오후 5 34 40](https://github.com/vercel/next.js/assets/81786662/52499900-38f2-40f5-94b9-2deaf816fc94)

### Additional information

```js
next.config

/**
 * Run `build` or `dev` with `SKIP_ENV_VALIDATION` to skip env validation.
 * This is especially useful for Docker builds.
 */
import withPWA from "next-pwa";
!process.env.SKIP_ENV_VALIDATION && (await import("./src/env.mjs"));

/** @type {import("next").NextConfig} */
const config = {
  reactStrictMode: true,
  headers: async () => [
    {
      source: "/myapp/:route",
      headers: [
        {
          key: "Cache-Control",
          value: "no-store",
        },
      ],
    },
  ],
  webpack(config) {
    config.module.rules.push({
      test: /\.svg$/i,
      issuer: /\.[jt]sx?$/,
      use: ["@svgr/webpack"],
    });

    return config;
  },

  /**
   * If you have the "experimental: { appDir: true }" setting enabled, then you
   * must comment the below `i18n` config out.
   *
   * @see https://github.com/vercel/next.js/issues/41980
   */
  i18n: {
    locales: ["en"],
    defaultLocale: "en",
  },
};
export default withPWA({
  dest: "public",
  disable: process.env.NODE_ENV === "development",
})(config);
```

### Example

.

**Comment by icyJoseph** [comment, marked as answer]
This sounds a bit to be a problem with next-pwa. The Service Worker gets installed on your browser, so I think, even if you disable PWA, the worker is still present.

Do this. 

1. Disable next-pwa
2. Clear the service workers (https://stackoverflow.com/a/41907900)

Try again. I suspect the service worker is the issue, because of the ⚙️ next to the request. That means that request is being served by a worker box. 

Also because you are not doing the necessary steps in `next.config.js`. Here I explain what you need to do: https://github.com/vercel/next.js/discussions/51454#discussioncomment-6218798

### Summary for visibility

`next-pwa`, by default has something called stale-while-revalidate, https://developer.chrome.com/docs/workbox/modules/workbox-strategies/#stale-while-revalidate, which for GSSP, is a problem.

You need to use https://developer.chrome.com/docs/workbox/modules/workbox-strategies/#network-first-network-falling-back-to-cache:

```js
/** @type {import('next').NextConfig} */
const runtimeCaching = require("next-pwa/cache");

const nextDataIndex = runtimeCaching.findIndex(
  (entry) => entry.options.cacheName === "next-data"
);

if (nextDataIndex !== -1) {
  runtimeCaching[nextDataIndex].handler = "NetworkFirst";
} else {
  throw new Error("Failed to find next-data object in runtime caching");
}

const withPWA = require("next-pwa")({
  dest: "public",
  runtimeCaching,
  register: true,
  skipWaiting: true,
  buildExcludes: [/middleware-manifest.json$/],
  disable: process.env.NODE_ENV === 'development'
})

module.exports = withPWA({
  reactStrictMode: false
});
```

**Reply by yeonhwan** [reply]
Thank you for answering!

Before I find your answer here, I was able to find your other answers from related questions (the one you linked above) and could fix the problem with that method. I also followed the method (disable pwa, clear service worker) as well but Im not sure about my application working because I changed service worker's settings or I reset the service worker.  

But now I'm having a few questions. Can you answer these please?
1.  PWA in next application, does it need third party packages to be achieved?
2. Does plain Next application use Service Worker for fetching / caching purposes?

Appreciate your efforts for community!

**Reply by icyJoseph** [reply]
> PWA in next application, does it need third party packages to be achieved?

If you mean, if Next.js has some built-in API to activate a worker, as far as I know, no. 

> Does plain Next application use Service Worker for fetching / caching purposes?

No it doesn't.

PWA and Service Workers are amazing, but historically, it has been kind of easy to get into trouble. I remember having apps that did not update as I wanted, or that still showed an old version several months after I pushed new stuff, because of a wonky Service Worker. 

You probably found on my answers the problem Kent C Dodds had... 

You can of course do it yourself, without using `next-pwa`, but you'd just end up recreating the same thing I think.

**Reply by rabieeali** [reply]
thank you!

but what are the solutions for me?
how can I use pwa in nextjs?

are there any alternative packages if I do not do it manually?

**Reply by icyJoseph** [reply]
Nothing blocks you from using `next-pwa`, you just have to do the change, so that GSSP is not opted into stale-while-revalidate. Read carefully the main comment. Notice how we switch pages using GSSP to `NetworkFirst`.

**Reply by danangtomo** [reply]
hi, @icyJoseph why when i import next-pwa/cache, will show "Could not find a declaration file for module 'next-pwa/cache'."

**Reply by icyJoseph** [reply]
Hi @danangtomo is your next config file a TS file? or something like that? Anyhow, I guess you need to also install, https://www.npmjs.com/package/@types/next-pwa

**Reply by danangtomo** [reply]
@icyJoseph  JS file. for my case i should refresh the page twice attemps for the data can be updated

and very often there shown "Uncaught (in promise) DOMException: Failed to execute 'put' on 'Cache': Unexpected internal error." in the console browser.
